src/Processors/Executors/PipelineExecutor.cpp

```diff
@@ -28,7 +35,8 @@ namespace ErrorCodes
 
 
 PipelineExecutor::PipelineExecutor(std::shared_ptr<Processors> & processors, QueryStatusPtr elem)
-    : process_list_element(std::move(elem))
+    : pool(CurrentMetrics::QueryPipelineExecutorThreads, CurrentMetrics::QueryPipelineExecutorThreadsActive)
```

Comment:
I don't like that thread pool settings are too implicit.
Default ctor use `getNumberOfPhysicalCPUCores` for `max_threads`, `max_free_threads`, and `queue_size`. Then, we change it into `pool.setMaxThreads(num_threads);`, and in case if `num_threads != getNumberOfPhysicalCPUCores` the result is not obvious (actually, looks like only `max_free_threads` could be > `max_threads` sometimes, which is ok; but I spend some time to verify it).
Let's maybe made it explicit.

Also, maybe make pool as `std::optional`, and init only when `num_threads > 1` in `executeImpl`.

Reply:
Agree, converted to std::unique_ptr (to make it possible to use forward declaration only)